### PR TITLE
Properly handle latin-1 encoding in file diffs

### LINKED ISSUE #48777: file.recurse fails on file contents encoding not utf-8
### Description of Issue/Question
Trying to distribute files with contents encoded in ISO-8859-1. However salt assumes systemdefault 
utf-8 and thus fails on illegal byte sequences when trying to merge updates.

### Setup
One saltmaster on Cent OS 7. One minion on Cent OS 7.

state file foo.sls:
```
webfoo:
  file.recurse:
    - name: /tmp/foo/
    - source: salt://foo/
    - user: root
    - group: root
    - file_mode: 444
    - dir_mode: 555
    - makedirs: True

```
[fileone.tar.gz](https://github.com/saltstack/salt/files/2231198/fileone.tar.gz)
[filetwo.tar.gz](https://github.com/saltstack/salt/files/2231199/filetwo.tar.gz)
### Steps to Reproduce Issue
Populate directory foo using fileone.tar (plain ascii I believe).
Distribute file using `salt '*' state.apply foo`
Update file on saltmaster using filetwo.tar (iso-8859-1 contents added).
Try distribute file using `salt '*' state.apply foo`
Get error:
          ID: webfoo
    Function: file.recurse
        Name: /tmp/foo/
      Result: False
     Comment: #### /tmp/foo/index.html ####
              Unable to manage file: 'utf8' codec can't decode byte 0xe4 in position 1: invalid continuation byte

### Versions Report
# salt --versions-report
Salt Version:
           Salt: 2018.3.2
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Jul 13 2018, 13:06:57)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: centos 7.5.1804 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-862.9.1.el7.x86_64
         system: Linux
        version: CentOS Linux 7.5.1804 Core

### PR DESCRIPTION
**Pull request by terminalmage**
This changes the encoding/decoding helpers to attempt latin-1 after initially trying UTF-8, which prevents a traceback when displaying diffs when a file with latin-1 encoding is updated using a file.managed state.

Resolves #48777

### CODE REVIEW

File: salt/modules/file.py [hunk 1]
```diff
@@ -1567,7 +1567,7 @@ def comment_line(path,
         check_perms(path, None, pre_user, pre_group, pre_mode)
 
     # Return a diff using the two dictionaries
-    return ''.join(difflib.unified_diff(orig_file, new_file))
+    return __salt__['stringutils.get_diff'](orig_file, new_file)
```
Comment (rallytime) [inline salt/modules/file.py]: Should this be `__utils__` instead of `__salt__`?

File: salt/modules/file.py [hunk 2]
```diff
@@ -2038,9 +2038,7 @@ def line(path, content=None, match=None, mode=None, location=None,
         if show_changes:
             with salt.utils.files.fopen(path, 'r') as fp_:
                 path_content = salt.utils.data.decode_list(fp_.read().splitlines(True))
-            changes_diff = ''.join(difflib.unified_diff(
-                path_content, body
-            ))
+            changes_diff = __salt__['stringutils.get_diff'](path_content, body)
```
Comment (rallytime) [inline salt/modules/file.py]: Same question here as above.

File: salt/modules/file.py [hunk 3]
```diff
@@ -2688,7 +2683,7 @@ def _add_content(linesep, lines=None, include_marker_start=True,
             )
 
     if block_found:
-        diff = ''.join(difflib.unified_diff(orig_file, new_file))
+        diff = __salt__['stringutils.get_diff'](orig_file, new_file)
```
Comment (rallytime) [inline salt/modules/file.py]: Here, too?

**Review by rallytime — COMMENTED** [review]
It seems that `__salt__` is being used instead of `__utils__` in a couple of places. There's some lint/test errors as well.

**Review by rallytime — APPROVED** [review]

**Review by damon-atkins — APPROVED** [review]

**Comment by rallytime** [comment]
@terminalmage This looks like the tests will need some mocking love, due to the change to `__utils__`.

**Comment by rallytime** [comment]
@terminalmage Looks like one last test failure to clean up here

https://jenkinsci.saltstack.com/job/pr-kitchen-ubuntu1604-py3/job/PR-48934/4/

**Pull request merged**